src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -2787,3 +2789,125 @@ def find_q(m, m4_fac, D):
+
+    - ``N`` -- integer; the order for which we seek an elliptic curve. Must be a
+      prime number.
+
+    OUTPUT: an elliptic curve `E/\mathbb F_p` of order ``N``
+
+    EXAMPLES::
+
+        sage: N = next_prime(int(b'sagemath'.hex(), 16))
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E
+        Elliptic Curve defined by y^2 = x^3 + 4757897140353078952*x +
+        1841350074072114366 over Finite Field of size 8314040074357871443
+        sage: E.order() == N
+        True
+
+    The execution time largely depends on the input::
+
+        sage: N = 125577861263605878504082476745517446213
+        sage: E = EllipticCurve_with_prime_order(N) # Takes ~1 second.
+        sage: E.order() == N
+        True
+
+    TESTS::
+
+        sage: for N in prime_range(3, 100):
+        ....:     E = EllipticCurve_with_prime_order(N)
+        ....:     assert E.order() == N
+
+        sage: N = 15175980689839334471
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E.order() == N
+        True
+
+        sage: N = next_prime(123456789)
+        sage: E = EllipticCurve_with_prime_order(N)
+        sage: E.order() == N
+        True
+
+        sage: N = 123456789
+        sage: E = EllipticCurve_with_prime_order(N)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+        sage: E = EllipticCurve_with_prime_order(0)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+        sage: E = EllipticCurve_with_prime_order(-7)
+        Traceback (most recent call last):
+        ...
+        ValueError: input order is not prime
+
+    .. NOTE::
+
+        Depending on the input, this function may run forever.
+
+    ALGORITHM: [BS2007]_, Algorithm 2.2
+    """
+    from sage.arith.misc import is_prime
+    from sage.combinat.subset import powerset
+    from sage.functions.other import ceil
+    from sage.misc.functional import symbolic_prod as product
+    from sage.quadratic_forms.binary_qf import BinaryQF
+    from sage.rings.fast_arith import prime_range
+    from sage.schemes.elliptic_curves.cm import hilbert_class_polynomial
+
+    if not is_prime(N):
+        raise ValueError("input order is not prime")
+
+    # The algorithm consists of multiple ‘search rounds’ for a suitable
+    # discriminant `D`, `r` defines the number of rounds. We expect this
+    # algorithm to terminate after a number of rounds that is polynomial in
+    # loglog N.
+    r = 0
+
+    while True:
+        # Iterating over the odd primes by chunks of size log(`N`).
+        S = prime_range(3, ceil((r + 1) * N.log()))
+
+        # Every possible products of distinct elements of `S`.
+        # There probably is a more optimal way to compute all possible products
+        # of elements of S than using a powerset. Here many multiplications are
+        # done multiple times.
+        for e in powerset(S):
+            D = product(e)
```

Comment:
In the paper, the algorithm takes a product of the `p^*` terms, not the `p` terms. Should there be a computation of the `p^*` terms somewhere? I also don't see where you do the square root mod `N` part of the algorithm.

Reply:
The square roots are used to solve the diophantine equation for `4 * N` and `BinaryQF` does it for us.

The `p*` are defined as `p* = (-1)^(p-1 / 2) * p`. Basically, it checks the parity of `p-1 / 2` to negate or not `p`. This is a shady part for me because this wouldn't work and it kept yielding non-discriminants for the Hilbert polynomial. I found a hack around that seems to work but maybe I'm wrong here.